File: apps/worker/worker/documents/renderer.py
from collections.abc import Mapping, Sequence
from io import BytesIO
from typing import cast

from docx import Document


def render_resume_docx(
	facts: Mapping[str, object], suggestions: Sequence[Mapping[str, object]]
) -> bytes:
	# The corrected document only reorganizes documented facts; it never adds
	# content the evaluation did not find in the source resume.
	document = Document()
	contact = mapping_of(facts.get("contact"))
	name = text_of(contact.get("name")) or "Resume"
	document.add_heading(name, level=0)
	contact_line = " · ".join(
		item
		for item in (
			text_of(contact.get("email")),
			text_of(contact.get("phone")),
			text_of(contact.get("location")),
		)
		if item
	)
	if contact_line:
		document.add_paragraph(contact_line)

	skills = entries_of(facts.get("skills"))
	if skills:
		document.add_heading("Skills", level=1)
		grouped: dict[str, list[str]] = {}
		for skill in skills:
			canonical = text_of(skill.get("canonicalName"))
			if not canonical:
				continue
			category = text_of(skill.get("category")) or "Other"
			grouped.setdefault(category, []).append(canonical)
		for category in sorted(grouped):
			document.add_paragraph(
				f"{category}: {', '.join(sorted(grouped[category], key=str.casefold))}"
			)

	employment = entries_of(facts.get("employment"))
	if employment:
		document.add_heading("Experience", level=1)
		for role in employment:
			title = " — ".join(
				item
				for item in (
					text_of(role.get("title")),
					text_of(role.get("employer")),
				)
				if item
			)
			start = text_of(role.get("startDate"))
			end = text_of(role.get("endDate"))
			if role.get("isCurrent") is True and start:
				dates = f"{start} to present"
			elif start and end:
				dates = f"{start} to {end}"
			else:
				dates = start or end or ""
			line = f"{title} ({dates})" if title and dates else title or dates or "Role"
			document.add_paragraph(line)

	education = entries_of(facts.get("education"))
	if education:
		document.add_heading("Education", level=1)
		for entry in education:
			parts = [
				item
				for item in (
					text_of(entry.get("degree")),
					text_of(entry.get("fieldOfStudy")),
					text_of(entry.get("institution")),
				)
				if item
			]
			if parts:
				document.add_paragraph(", ".join(parts))

	certifications = entries_of(facts.get("certifications"))
	if certifications:
		document.add_heading("Certifications", level=1)
		for entry in certifications:
			parts = [
				item
				for item in (
					text_of(entry.get("name")),
					text_of(entry.get("issuer")),
				)
				if item
			]
			if parts:
				document.add_paragraph(" — ".join(parts))

	if suggestions:
		document.add_heading("Improvement notes", level=1)
		for suggestion in suggestions:
			title = text_of(suggestion.get("title"))
			detail = text_of(suggestion.get("detail"))
			line = f"{title}: {detail}" if title and detail else title or detail
			if line:
				document.add_paragraph(line, style="List Bullet")

	buffer = BytesIO()
	document.save(buffer)
	return buffer.getvalue()


def mapping_of(value: object) -> Mapping[str, object]:
	if isinstance(value, Mapping):
		return cast(Mapping[str, object], value)
	return {}


def entries_of(value: object) -> list[Mapping[str, object]]:
	if not isinstance(value, list):
		return []
	return [
		cast(Mapping[str, object], item)
		for item in cast(list[object], value)
		if isinstance(item, Mapping)
	]


def text_of(value: object) -> str | None:
	return value if isinstance(value, str) and value.strip() else None

File: apps/worker-new/src/documents/renderer.ts
import { Document, HeadingLevel, Packer, Paragraph } from "docx";

const MAX_RENDERED_TEXT_CHARACTERS = 2000;

type Entry = Record<string, unknown>;

const mappingOf = (value: unknown): Entry =>
	typeof value === "object" && value !== null && !Array.isArray(value) ? (value as Entry) : {};

const entriesOf = (value: unknown): Entry[] =>
	Array.isArray(value) ? value.filter((item): item is Entry => typeof item === "object" && item !== null && !Array.isArray(item)) : [];

const textOf = (value: unknown): string | null => {
	if (typeof value !== "string") return null;
	const cleaned = [...value].filter((character) => character === "\t" || character === "\n" || character === "\r" || (character.codePointAt(0) ?? 0) >= 0x20).join("").trim();
	if (!cleaned) return null;
	return cleaned.slice(0, MAX_RENDERED_TEXT_CHARACTERS);
};

export const renderResumeDocx = async (facts: Readonly<Record<string, unknown>>, suggestions: readonly Entry[]): Promise<Uint8Array> => {
	const contact = mappingOf(facts["contact"]);
	const name = textOf(contact["name"]) ?? "Resume";
	const children: Paragraph[] = [new Paragraph({ text: name, heading: HeadingLevel.TITLE })];
	const contactLine = [textOf(contact["email"]), textOf(contact["phone"]), textOf(contact["location"])].filter((item): item is string => Boolean(item)).join(" · ");
	if (contactLine) children.push(new Paragraph(contactLine));

	const skills = entriesOf(facts["skills"]);
	if (skills.length) {
		children.push(new Paragraph({ text: "Skills", heading: HeadingLevel.HEADING_1 }));
		const grouped = new Map<string, string[]>();
		for (const skill of skills) {
			const canonical = textOf(skill["canonicalName"]);
			if (!canonical) continue;
			const category = textOf(skill["category"]) ?? "Other";
			grouped.set(category, [...(grouped.get(category) ?? []), canonical]);
		}
		for (const category of [...grouped.keys()].sort()) {
			children.push(new Paragraph(`${category}: ${[...(grouped.get(category) ?? [])].sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase())).join(", ")}`));
		}
	}

	const employment = entriesOf(facts["employment"]);
	if (employment.length) {
		children.push(new Paragraph({ text: "Experience", heading: HeadingLevel.HEADING_1 }));
		for (const role of employment) {
			const title = [textOf(role["title"]), textOf(role["employer"])].filter((item): item is string => Boolean(item)).join(" — ");
			const start = textOf(role["startDate"]);
			const end = textOf(role["endDate"]);
			const dates = role["isCurrent"] === true && start ? `${start} to present` : start && end ? `${start} to ${end}` : start ?? end ?? "";
			const line = title && dates ? `${title} (${dates})` : title || dates || "Role";
			children.push(new Paragraph(line));
		}
	}

	const education = entriesOf(facts["education"]);
	if (education.length) {
		children.push(new Paragraph({ text: "Education", heading: HeadingLevel.HEADING_1 }));
		for (const entry of education) {
			const parts = [textOf(entry["degree"]), textOf(entry["fieldOfStudy"]), textOf(entry["institution"])].filter((item): item is string => Boolean(item));
			if (parts.length) children.push(new Paragraph(parts.join(", ")));
		}
	}

	const certifications = entriesOf(facts["certifications"]);
	if (certifications.length) {
		children.push(new Paragraph({ text: "Certifications", heading: HeadingLevel.HEADING_1 }));
		for (const entry of certifications) {
			const parts = [textOf(entry["name"]), textOf(entry["issuer"])].filter((item): item is string => Boolean(item));
			if (parts.length) children.push(new Paragraph(parts.join(" — ")));
		}
	}

	if (suggestions.length) {
		children.push(new Paragraph({ text: "Improvement notes", heading: HeadingLevel.HEADING_1 }));
		for (const suggestion of suggestions) {
			const title = textOf(suggestion["title"]);
			const detail = textOf(suggestion["detail"]);
			const line = title && detail ? `${title}: ${detail}` : title ?? detail;
			if (line) children.push(new Paragraph({ text: line, bullet: { level: 0 } }));
		}
	}

	return new Uint8Array(await Packer.toBuffer(new Document({ sections: [{ children }] })));
};
